### PR TITLE
203: Request throttling for ODK Collect endpoints

### PR DESCRIPTION
**Pull request by md5**
This PR implements the request throttling portion of issue 203 for the ODK Collect endpoints.

I haven't yet made it configurable, but that would involve using a `proc` for `:limit` (to ensure that the config has been fully loaded). Let me know if you actually want this to be configurable.

I also added a spec.

One thing I wasn't sure about was whether to make `direct_auth?` match paths without `.xml`. There are tests in `basic_authentication_spec.rb` that aren't using `.xml`, so I'm not sure what's actually being requested by ODK Collect. I also recall @hooverlunch mentioning that there is a potentially conflicting path for `/forms/:id`.

### CODE REVIEW

**Comment by hooverlunch** [comment]
Good question about paths. 

The only path that will have a .xml ending is /m/mission1/forms/123.xml

All the other direct auth paths will not.

Another way to distinguish them perhaps is by the fact that they have no locale.

If you can think of some better way of organizing this feel free. It's just that we have no control over the /formList or /submission. Anything else we can change.

File: config/initializers/rack-attack.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
+class Rack::Attack::Request
+  def direct_auth?
+    if path =~ %r{^/m/[a-z][a-z0-9]*/(.*)$}
+      subpath = $1
+
+      return true if subpath == 'formList.xml'
+      return true if subpath =~ %r{^forms/([^/]+)\.xml$}
+      return true if subpath =~ %r{^forms/([^/]+)/manifest\.xml$}
+      return true if subpath =~ %r{^forms/([^/]+)/itemsets\.csv$}
+      return true if subpath == 'submission.xml'
```
Comment (hooverlunch) [inline config/initializers/rack-attack.rb]: So remove .xml from all but line 7.

File: spec/requests/request_throttling_spec.rb [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  let(:limit) { 30 }
+
+  before(:each) do
+    Rack::Attack.cache.store = ActiveSupport::Cache::MemoryStore.new
+  end
+
+  context 'with the same ip address' do
+    it 'should not apply to requests below the limit' do
+      limit.times do
+        get "/m/#{get_mission.compact_name}/formList.xml"
```
Comment (hooverlunch) [inline spec/requests/request_throttling_spec.rb]: Should change the .xmls down here too.

File: spec/requests/request_throttling_spec.rb [hunk 3]
```diff
@@ -0,0 +1,42 @@
+require 'spec_helper'
+
+describe 'throttling for xml requests' do
+
+  let(:limit) { 30 }
```
Comment (hooverlunch) [inline spec/requests/request_throttling_spec.rb]: I'm concerned about how long this test will take to run. Could we stub the limit to like 3 or 4 just for the test case?

File: spec/requests/request_throttling_spec.rb [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+    it 'should not apply' do
+      (limit * 2).times do |i|
+        get "/m/#{get_mission.compact_name}/formList.xml", nil, { REMOTE_ADDR: remote_addrs[i] }
+        assert_response :unauthorized
+      end
+    end
+  end
+
+end
```
Comment (hooverlunch) [inline spec/requests/request_throttling_spec.rb]: Good job on the test.

File: config/routes.rb [hunk 5]
```diff
@@ -131,6 +131,11 @@
     get '/sms/submit' => 'sms#create'
```
Comment (hooverlunch) [inline config/routes.rb]: Can we move the SMS routes into a separate scope block, then your comment below can apply to the whole ODK scope block.

The SMS stuff is going to change soon too, with the token ticket that is upcoming.

Also we could perhaps put direct_auth: true in the defaults for the whole block to simplify things. The condition in authentication.rb would still work even for the the no_auth one. We could consider no_auth a form of direct_auth, if you catch my drift.

**Comment by hooverlunch** [comment]
Nice work. Just a few small things and you can merge.

File: spec/requests/request_throttling_spec.rb (hunk 3 above)
Comment (md5) [inline spec/requests/request_throttling_spec.rb]: Did you try running it? It runs in less than a second.

```
$ bundle exec rspec spec/requests/request_throttling_spec.rb

Randomized with seed 35489

throttling for xml requests
  with the same ip address
    should not apply to requests below the limit
    should apply to requests above the limit with response code 429
  with different ip addresses
    should not apply

Finished in 5.18 seconds (files took 4.3 seconds to load)
3 examples, 0 failures

Randomized with seed 35489
```

Comment (md5) [inline spec/requests/request_throttling_spec.rb]: Also, the limit here has to match the limit in `rack-attack.rb` for the test to work right. If I were to do the `proc` for `:limit`, both values could be pulled from config.

**Comment by md5** [comment]
> The only path that will have a .xml ending is /m/mission1/forms/123.xml
>
> All the other direct auth paths will not.

I'll update the code to reflect that.

> Another way to distinguish them perhaps is by the fact that they have no locale.

Since this middleware runs in front of the Rails stack, I don't think we have access to a locale yet.

**Comment by md5** [comment]
Oh, wait, you mean the locale in the URL. I'll think about whether that can simplify this.

**Comment by hooverlunch** [comment]
You already take this into account via your regexp I believe. So all is well on this front I think.

**Comment by md5** [comment]
I've updated this to include a negative test to make sure SMS requests aren't being throttled by `Rack::Attack`.

This has ended up being a bunch of commits, so let me know if you'd like me to squash them.

**Comment by hooverlunch** [comment]
Looks great!

**Pull request merged**